### ISSUE TITLE
memo() example

### ISSUE THREAD

**Issue by dakom** [WebReflection/augmentor#7]
See: https://codepen.io/dakom/pen/QzRmar?editors=0010

Not sure if it can be built off `useMemo` or if we need a new primitive

**Comment by WebReflection** [comment]
so, not sure about the whole test, but during some change I've lost `useMemo` returned `value` so it was plain broken.

**Comment by WebReflection** [comment]
to be honest, I'm a bit disappointed all the coverage was fine, I need to also improve assertions there or this stuff might have unnoticed regressions .. anyway, once unpkg picks up the latest neverland, we'll see if anything is better at all.

**Comment by dakom** [comment]
OK cool - I changed the codepen to use just this:

```
const memo = component(child => props => {
  return useMemo(() => child(props), [props]);
});
```

I see it's not crashing anymore which is great! Though it's not a valid `memo()`

However, I think this might match React and I'm guessing it's why React provides both `memo()` and `useMemo()`

Also, is it always supposed to be just a pure thunk, i.e. there's never a useful argument passed as the `() =>` part?

**Comment by dakom** [comment]
Not that React needs to be like the source of truth or anything, but I think it's a good idea to at least look at it as a reference

**Comment by dakom** [comment]
Yup! Changing React to the same approach to memo gives the same results: https://codepen.io/dakom/pen/REmvzw

**Comment by WebReflection** [comment]
> Though it's not a valid memo()

I've no idea what is `memo` about. Any documentation ?

> there's never a useful argument passed as the () => part?

the function in memo is used so that it's invoked only if it's needed. AFAIK no arguments are passed so it should be aligned.

Yet, I've no idea what is `memo`.

**Comment by dakom** [comment]
I emphasized the part that would help with this test:

> If your function component renders the same result given the same props, you can wrap it in a call to React.memo for a performance boost in some cases by memoizing the result. **This means that React will skip rendering the component, and reuse the last rendered result.**
> 
> By default it will only shallowly compare complex objects in the props object. If you want control over the comparison, you can also provide a custom comparison function as the second argument.
> 
> This method only exists as a performance optimization. Do not rely on it to “prevent” a render, as this can lead to bugs.

- https://reactjs.org/docs/react-api.html#reactmemo

Though tbh, the point of that context test was... for context, not this funky memo() function... not sure how necessary this is. I think I'll ask on the React github to learn a little bit more about this... "can lead to bugs" sounds horrible...

**Comment by dakom** [comment]
For reference: https://github.com/facebook/react/issues/14616

**Comment by WebReflection** [comment]
So, I've reverted the automatic assignment of the component since it's clearly broken due various reasons, as example that when conditional, i.e. behind a memo, the component index/invoke completely arbitrary.

I need to better think if it's even possible to have the automatic index at this point (I believe it is) but at least I want to be sure that everything else in augmentor works as expected, as it seems to be the case here with this memo, which is now showing same result (unkeyed but that's another story).

- - -

P.S. another thing I've noticed you use `component(...)` for any callback. You don't need to do that, you can use `useMemo` or anything else so that components should be **only** those you render, no need to augment everything that is a function.

As example, these are just fine.
```js
const withTick = Component => props => {
  const tick = useContext(TickContext);
  return Component(Object.assign({}, props, {tick}));
};

const memo = Component => props => useMemo(
  () => Component(props),
  Object.keys(props).map(key => props[key])
);
```

**Comment by dakom** [comment]
Cool, thanks for the heads up!

Btw I've got to make a decision between Neverland and React for an upcoming work project, and I think as much as I want it to be Neverland I've got to go with React now until Neverland is more stable... but I actually think it won't be too hard to switch later as long as you follow the same hooks API.

Really enjoyed writing in template literals and I like what you've got going on - looking forward to trying it out again a bit further down the line.

Is there a mailing list or something where I can get updates on major releases? I do want to stay in the loop of where things are at even if I'm not using it daily.
